soc/usbexecd.c

```diff
@@ -0,0 +1,1029 @@
+#define info(args...)   syslog(LOG_INFO, args)
+#define error(args...)  syslog(LOG_ERR, args)
+#define fatal(args...) ({ \
+    syslog (LOG_CRIT, args); \
+    exit(1); \
+})
+
+int sock = 0;
+int epfd   = 0;
+int sigfd  = 0;
+
+#define PIPE_BUF 4096
+#define MAX_CTRL_ARGS 255
+#define MAX_WRITE_LEN 255
+
+enum Roles {
+    ROLE_CTRL = 0,
+    ROLE_STDIN = 1,
+    ROLE_STDOUT = 2,
+    ROLE_STDERR = 3,
+};
+
+typedef struct {
+    // The id set by the CLI
+    uint8_t id;
+    // The stream role (see Roles enum)
+    uint8_t role;
+    // The epoll events to use as triggers
+    int events;
+    // Whether or not this stream has been requested to close
+    bool eof;
+    // The amount of bytes the other endpoint can receive
+    int credit;
+    // The file descriptor that this pipe buffer reads/writes data to
+    int fd;
+    // The index of the first readable byte of the ring buffer
+    int startpos;
+    // The index of the first writable byte of the ring buffer
+    int endpos;
+    // The number of 'active' bytes
+    int bufcount;
+    // The internal buffer used for back pressure
+    char buffer[PIPE_BUF];
+} pipebuf_t;
+
+typedef struct {
+    int pid;
+    pipebuf_t ctrl;
+    pipebuf_t stdin;
+    pipebuf_t stdout;
+    pipebuf_t stderr;
+} procinfo_t;
+
+#define MAX_COMMAND_LEN 1024
+#define N_PROC 256
+procinfo_t* processes[N_PROC];
+
+void child(int ctrl, int stdin, int stdout, int stderr);
+void pipebuf_out_ack(pipebuf_t* pb, size_t acksize);
+void modify_pipebuf_epoll(pipebuf_t* pb, int operation);
+void add_pipebuf_epoll(pipebuf_t* pb);
+void delete_pipebuf_epoll(pipebuf_t* pb);
+void pipebuf_interal_write(pipebuf_t* pb, uint8_t *data);
+int write_from_pipebuf(pipebuf_t *pb, int fd, int len);
+void pipebuf_out_to_internal_buffer(pipebuf_t* pb, int len);
+void pipebuf_out_is_writable(pipebuf_t* pb);
+void pipebuf_common_debug(pipebuf_t *pb, const char * str);
+
+// Helper function to write a packet header to the domain socket
+// Param: cmd - which command in the Commands enum to send
+// Param: id - which process is sending this
+// Param: arg - an argument for the command (like ACK)
+// Param len - the length of the data being sent following this packet
+void send_header(uint8_t cmd, uint8_t id, uint8_t arg, uint8_t len) {
+    // Create the packet
+    uint8_t buf[4] = {cmd, id, arg, len};
+    // Write the packet to the socket
+    int r = write(sock, buf, sizeof(buf));
+    // Ensure it wrote all of the bytes
+    if (r != sizeof(buf)) {
+        fatal("send_header failed: %s", strerror(errno));
+    }
+}
+
+int read_until(int fd, void *buf, int len) {
+
+    int total_read = 0, single_read = 0;
+
+    while (total_read < len) {
+
+        // Read from a file descriptor into another buffer
+        single_read = read(fd, &(((uint8_t *)buf)[total_read]), len - total_read);
+
+        // If we read nothing and the socket would have blocked
+        if (single_read < 0 && (errno == EAGAIN || errno == EWOULDBLOCK)) {
```

Comment:
If you use this function on a nonblocking fd and it returns EWOULDBLOCK, you're going to spin in this loop until it's readable, wasting a bunch of CPU time.